File: jun1/src/u8set.rs
use std::ops::{BitAnd, BitAndAssign, BitOr, BitOrAssign};
use crate::bitset256::BitSet256;

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct U8Set {
    bitset: BitSet256,
}

impl U8Set {
    pub fn insert(&mut self, value: u8) -> bool {
        if self.contains(value) {
            false
        } else {
            self.bitset.set_bit(value);
            true
        }
    }

    pub fn remove(&mut self, value: u8) -> bool {
        if !self.contains(value) {
            false
        } else {
            self.bitset.clear_bit(value);
            true
        }
    }

    pub fn update(&mut self, other: &U8Set) {
        self.bitset.update(&other.bitset);
    }

    pub fn contains(&self, value: impl Into<u8>) -> bool {
        self.bitset.is_set(value.into())
    }

    pub fn len(&self) -> usize {
        self.bitset.len()
    }

    pub fn is_empty(&self) -> bool {
        self.bitset.is_empty()
    }

    pub fn clear(&mut self) {
        self.bitset.clear();
    }

    pub fn all() -> Self {
        U8Set { bitset: BitSet256::all() }
    }

    pub fn none() -> Self {
        U8Set { bitset: BitSet256::none() }
    }

    pub fn from_chars(chars: &str) -> Self {
        let mut result = Self::none();
        for c in chars.chars() {
            result.insert(c as u8);
        }
        result
    }

    pub fn from_match_fn<F>(f: F) -> Self
    where
        F: Fn(u8) -> bool,
    {
        let mut result = Self::none();
        for i in 0..=255 {
            if f(i) {
                result.insert(i);
            }
        }
        result
    }

    pub fn iter(&self) -> impl Iterator<Item = u8> + '_ {
        (0..=255).filter(move |&i| self.contains(i))
    }
}

impl BitOr for &U8Set {
    type Output = U8Set;

    fn bitor(self, other: &U8Set) -> U8Set {
        U8Set { bitset: self.bitset.clone() | other.bitset.clone() }
    }
}

impl BitAnd for &U8Set {
    type Output = U8Set;

    fn bitand(self, other: &U8Set) -> U8Set {
        U8Set {
            bitset: self.bitset.clone() & other.bitset.clone(),
        }
    }
}

impl BitOr for U8Set {
    type Output = U8Set;

    fn bitor(self, other: U8Set) -> U8Set {
        &self | &other
    }
}

impl BitAnd for U8Set {
    type Output = U8Set;

    fn bitand(self, other: U8Set) -> U8Set {
        &self & &other
    }
}

impl BitOrAssign for U8Set {
    fn bitor_assign(&mut self, other: U8Set) {
        self.update(&other);
    }
}

impl BitAndAssign for U8Set {
    fn bitand_assign(&mut self, other: U8Set) {
        self.bitset &= other.bitset;
    }
}

impl std::fmt::Debug for U8Set {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "U8Set({:?})", self.iter().map(|i| i as char).collect::<Vec<_>>())
    }
}

impl std::fmt::Display for U8Set {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_u8set() {
        let mut set = U8Set::none();
        assert!(set.insert(b'a'));
        assert!(set.insert(b'b'));
        assert!(!set.insert(b'a'));
        assert!(set.contains(b'a'));
        assert!(set.contains(b'b'));
        assert!(!set.contains(b'c'));
        assert_eq!(set.len(), 2);
        assert!(set.remove(b'a'));
        assert!(!set.remove(b'c'));
        assert_eq!(set.len(), 1);
        assert!(!set.is_empty());
        set.clear();
        assert!(set.is_empty());

        let set1 = U8Set::from_chars("abc");
        let set2 = U8Set::from_chars("bcd");
        let union = &set1 | &set2;
        let intersection = &set1 & &set2;
        assert_eq!(union.len(), 4);
        assert_eq!(intersection.len(), 2);

        let even_set = U8Set::from_match_fn(|x| x % 2 == 0);
        assert!(even_set.contains(0));
        assert!(even_set.contains(2));
        assert!(!even_set.contains(1));
        assert_eq!(even_set.len(), 128);
    }
}

File: jun1/src/u8set.py
from __future__ import annotations

from typing import Iterator

from .bitset256 import BitSet256


class U8Set:
    def __init__(self, bitset: Optional[BitSet256] = None) -> None:
        self.bitset = bitset or BitSet256()

    @staticmethod
    def from_char(char: str) -> U8Set:
        return U8Set.from_chars(char)

    @staticmethod
    def from_range(start: int, end: int) -> U8Set:
        u8set = U8Set()
        for i in range(start, end + 1):
            u8set.insert(i)
        return u8set

    def insert(self, value: int) -> bool:
        if self.contains(value):
            return False
        self.bitset.set_bit(value)
        return True

    def remove(self, value: int) -> bool:
        if not self.contains(value):
            return False
        self.bitset.clear_bit(value)
        return True

    def update(self, other: U8Set) -> None:
        self.bitset.update(other.bitset)

    def contains(self, value: int) -> bool:
        return self.bitset.is_set(value)

    def __len__(self) -> int:
        return len(self.bitset)

    def is_empty(self) -> bool:
        return self.bitset.is_empty()

    def clear(self) -> None:
        self.bitset.clear()

    @staticmethod
    def all() -> U8Set:
        return U8Set(BitSet256.all())

    @staticmethod
    def none() -> U8Set:
        return U8Set()

    @staticmethod
    def from_chars(chars: str) -> U8Set:
        result = U8Set()
        for char in chars:
            result.insert(ord(char))
        return result

    def __iter__(self) -> Iterator[int]:
        for i in range(256):
            if self.contains(i):
                yield i

    def __or__(self, other: U8Set) -> U8Set:
        return U8Set(self.bitset | other.bitset)

    def __and__(self, other: U8Set) -> U8Set:
        return U8Set(self.bitset & other.bitset)

    def __ior__(self, other: U8Set) -> U8Set:
        self.bitset |= other.bitset
        return self

    def __iand__(self, other: U8Set) -> U8Set:
        self.bitset &= other.bitset
        return self

    def __repr__(self) -> str:
        return f"U8Set({', '.join(chr(i) for i in self)})"
